src/Sulu/Component/Content/Document/Subscriber/RouteGeneratingSubscriber.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Document\Subscriber;
+
+use Sulu\Bundle\ContentBundle\Document\HomeDocument;
+use Sulu\Bundle\ContentBundle\Document\RouteDocument;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\DocumentInspector;
+use Sulu\Component\Content\Document\Behavior\ResourceSegmentBehavior;
+use Sulu\Component\Content\Document\Behavior\RouteBehavior;
+use Sulu\Component\Content\Document\Behavior\WebspaceBehavior;
+use Sulu\Component\DocumentManager\DocumentManagerInterface;
+use Sulu\Component\DocumentManager\Event\HydrateEvent;
+use Sulu\Component\DocumentManager\Event\PersistEvent;
+use Sulu\Component\DocumentManager\Event\RemoveEvent;
+use Sulu\Component\DocumentManager\Events;
+use Sulu\Component\PHPCR\SessionManager\SessionManagerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Behavior for route (sulu:path) documents.
+ */
+class RouteGeneratingSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
+     * @var DocumentInspector
+     */
+    private $documentInspector;
+
+    /**
+     * @var SessionManagerInterface
+     */
+    private $sessionManager;
+
+    public function __construct(
+        DocumentManagerInterface $documentManager,
+        DocumentInspector $documentInspector,
+        SessionManagerInterface $sessionManager
+    ) {
+        $this->documentManager = $documentManager;
+        $this->documentInspector = $documentInspector;
+        $this->sessionManager = $sessionManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            // must be exectued before the TargetSubscriber
+            Events::PERSIST => ['handlePersist', 5],
+            Events::REMOVE => ['handleRemove', 550],
+        ];
+    }
+
+    /**
+     * Updates the route for the given document and creates history routes if necessary.
+     *
+     * @param PersistEvent $event
+     */
+    public function handlePersist(PersistEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof RouteBehavior) {
+            return;
+        }
+
+        // Do not generate route when "safe"
+        // TODO: Ligimitize this or cast it out.
+        if ($event->getOption('safe') === true) {
```

Comment:
This option allows us to not execute the route generation .. and consequently allows the synchronization to work in both directions.

The reason this class was split was a different solution to the same problem.